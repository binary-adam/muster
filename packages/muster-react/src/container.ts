import { deprecated } from '@dws/muster';
import identity from 'lodash/identity';
import {
  ContainerComponentFactory,
  ContainerConfiguration,
  Props,
  ScopeTransformer,
} from './container-types';
import createContainerFactory from './utils/create-container-factory';

const showDataDeprecationWarning = deprecated({ old: 'data', new: 'graph' });
const showCreateContainerDeprecationWarning = deprecated({
  old: 'createContainer',
  new: 'container',
});

function getComponentGraph<PF, P extends Props<PF>, RF, R extends Props<RF>>(
  configuration: ContainerConfiguration<PF, P, RF, R>,
): any {
  if (configuration.data) {
    showDataDeprecationWarning();
    return configuration.data;
  }
  return configuration.graph;
}

/**
 * Creates a new Muster React container.
 * @param configuration
 * @param transform
 * @deprecated
 */
export function createContainer<PF, P extends Props<PF>, RF, R extends Props<RF>>(
  configuration: ContainerConfiguration<PF, P, RF, R>,
  transform: ScopeTransformer = identity,
): ContainerComponentFactory<PF, P, RF, R> {
  showCreateContainerDeprecationWarning();
  return container(configuration, transform);
}

export function container<PF, P extends Props<PF>, RF, R extends Props<RF>>(
  configuration: ContainerConfiguration<PF, P, RF, R>,
  transform: ScopeTransformer = identity,
): ContainerComponentFactory<PF, P, RF, R> {
  return createContainerFactory(
    {
      events: configuration.events,
      graph: getComponentGraph(configuration),
      props: configuration.props,
      relaxPropsValidation: configuration.relaxPropsValidation,
      renderError: configuration.renderError,
      renderLoading: configuration.renderLoading,
      require: configuration.require,
    },
    transform,
  );
}
