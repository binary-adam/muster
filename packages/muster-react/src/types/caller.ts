import { createMatcher, getInvalidTypeError, isMatcher, Matcher, types } from '@dws/muster';

export interface CallerOptions<N, A> {
  args: A;
  name: N;
}

export type CallerMatcher<N, A> = Matcher<Function, CallerOptions<N, A>>;

/**
 * Creates a matcher used as part of the `props` section of the [[container]] and [[simpleContainer]]. This matcher
 * informs muster-react that a given property should be loaded from the graph as a function. The returned function
 * is an asynchronous function that returns a promise when called. Calling it will cause a matching `callable` node in the
 * graph to be called with the help of a `call()` node.
 *
 * In order for the function returned from the graph to work correctly the node in your graph must support `call`
 * operation. Examples of nodes that support `call` operation: [[action]], [[fn]], [[placeholder]].
 * If the node targeted by the [[caller]] does not support the `call` operation then calling the returned function will
 * result in the promise being rejected with an error.
 *
 *
 * @example **Call an action from the graph**
 * ```jsx
 * import { action, container, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     myAction: action(() => {
 *       console.log('myAction was called');
 *     }),
 *   },
 *   props: {
 *     myAction: propTypes.caller(),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ myAction }) => (
 *   <button onClick={myAction}>Click me</button>
 * ));
 *
 * // Console output (after clicking the button):
 * // myAction was called
 * ```
 * This example shows how to use the [[caller]] prop type to be able to call an action defined in the local container
 * graph. The code above ignores the fact that the `myAction` is an asynchronous function that returns a promise, and
 * just calls it every time the button is clicked. In most cases this is how a muster action should be called from your
 * components.
 *
 *
 * @example **Call an action from the graph, and read the return value**
 * ```js
 * import { action, container, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     myAction: action(() => {
 *       console.log('myAction was called');
 *       return 'some value from myAction';
 *     }),
 *   },
 *   props: {
 *     myAction: propTypes.caller(),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ myAction }) => {
 *   async function onButtonClick() {
 *     const result = await myAction();
 *     console.log('Result: ', result);
 *   }
 *   return (
 *     <button onClick={onButtonClick}>Click me</button>
 *   );
 * });
 *
 * // Console output (after clicking the button):
 * // myAction was called
 * // Result: some value from myAction
 * ```
 * This example shows how to utilise the asynchronous nature of the functions returned from the graph.
 *
 *
 * @example **Call an action with arguments**
 * ```js
 * import { action, container, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     myAction: action((name) => {
 *       console.log('myAction was called with name: ', name);
 *       return `Hello, ${name}`;
 *     }),
 *   },
 *   props: {
 *     myAction: propTypes.caller(),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ myAction }) => {
 *   async function onButtonClick() {
 *     const result = await myAction('Bob');
 *     console.log('Result: ', result);
 *   }
 *   return (
 *     <button onClick={onButtonClick}>Click me</button>
 *   );
 * });
 *
 * // Console output (after clicking the button):
 * // myAction was called with name: Bob
 * // Result: Hello, Bob
 * ```
 * This example shows how to pass an argument to an action. There is no limit to the number of arguments that can be
 * passed to the action, just like in a normal JavaScript function.
 *
 *
 * @example **Create a caller with a different name than the graph action**
 * ```js
 * import { action, container, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     someActionNameYouDontLike: action(() => {
 *       console.log('someActionNameYouDontLike was called');
 *     }),
 *   },
 *   props: {
 *     myAction: propTypes.caller('someActionNameYouDontLike'),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ myAction }) => (
 *   <button onClick={myAction}>Click me</button>
 * ));
 *
 * // Console output (after clicking the button):
 * // someActionNameYouDontLike was called
 * ```
 * The [[caller]] prop type also enables aliasing actions. In this example the action in the graph was named
 * `someActionNameYouDontLike`, which for some reason we don't find to our taste when defining props. With the
 * help of [[caller]] you can define the name of the prop to be completely different than the name of the action
 * defined in the graph.
 *
 *
 * @example **Validate types of arguments**
 * ```js
 * import { action, container, propTypes, types } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     myAction: action((name) => {
 *       console.log('myAction was called with name: ', name);
 *       return `Hello, ${name}`;
 *     }),
 *   },
 *   props: {
 *     myAction: propTypes.caller([types.string]),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ myAction }) => {
 *   async function onButtonClick() {
 *     console.log('Calling `myAction` with a string argument');
 *     try {
 *       const result = await myAction('Bob');
 *       console.log('Result: ', result);
 *     } catch (ex) {
 *       console.log('Didn\'t work');
 *     }
 *
 *     console.log('Calling `myAction` with a number argument');
 *     try {
 *       const result = await myAction(123);
 *       console.log('Result: ', result);
 *     } catch (ex) {
 *       console.log('Didn\'t work');
 *     }
 *   }
 *   return (
 *     <button onClick={onButtonClick}>Click me</button>
 *   );
 * });
 *
 * // Console output (after clicking the button):
 * // Calling `myAction` with a string argument
 * // myAction was called with name: Bob
 * // Result: Hello, Bob
 * // Calling `myAction` with a number argument
 * // Didn't work
 * ```
 * This example shows how to add validation to the type of arguments accepted by the callable function returned from the
 * graph. In the first `try` block of the `onButtonClick` we're calling the `myAction` with a string 'Bob', which returns
 * a correct value. In the next block we try our luck by passing a numeric value `123`, but muster-react catches us
 * before that argument ends up in the body of the `myAction`.
 */
export function caller(): CallerMatcher<undefined, undefined>;
export function caller(name: string): CallerMatcher<string, undefined>;
export function caller<PT, PP, P extends Matcher<PT, PP>, A extends Array<P>>(
  args: A,
): CallerMatcher<undefined, A>;
export function caller<PT, PP, P extends Matcher<PT, PP>, A extends Array<P>>(
  name: string,
  args: A,
): CallerMatcher<string, A>;
export function caller<PT, PP, P extends Matcher<PT, PP>, A extends Array<P>>(
  ...args: Array<string | A>
): CallerMatcher<string | undefined, A | undefined> {
  const options: CallerOptions<string | undefined, A | undefined> = {
    args: undefined,
    name: undefined,
  };
  // function caller<PT, PP, P extends Matcher<PT, PP>>(name: string, args: Array<P>): Matcher<never, CallerOptions<string, Array<P>>>
  if (args.length === 2 && Array.isArray(args[1]) && (args[1] as Array<any>).every(isMatcher)) {
    const [name, argTypes] = args as [string, A];
    options.name = name;
    options.args = argTypes;
  }
  // function caller(name: string): Matcher<never, CallerOptions<string, undefined>>
  else if (args.length === 1 && typeof args[0] === 'string') {
    const [name] = args as [string];
    options.name = name;
  }
  // function caller<PT, PP, P extends Matcher<PT, PP>>(args: Array<P>): Matcher<never, CallerOptions<undefined, Array<P>>>
  else if (
    args.length === 1 &&
    Array.isArray(args[0]) &&
    (args[0] as Array<any>).every(isMatcher)
  ) {
    const [argTypes] = args as [A];
    options.args = argTypes;
  } else if (args.length !== 0) {
    throw getInvalidTypeError('Invalid parameters supplied to the caller().', {
      expected: ['string', 'Array<Matcher>'],
      received: args,
    });
  }
  const matcher = createMatcher<Function, CallerOptions<string | undefined, A | undefined>>(
    'caller',
    (value: any) => types.func(value),
    options,
  );
  matcher.metadata.type = caller;
  return matcher;
}

export function isCallerMatcher(value: any): value is CallerMatcher<any, any> {
  return isMatcher(value) && value.metadata.type === caller;
}

export type CallerArgumentMatcher<N, A> = Matcher<Array<any>, CallerMatcher<N, A>>;

export function callerArguments<
  PT,
  PP,
  P extends Matcher<PT, PP>,
  A extends Array<P> | undefined,
  N
>(callerMatcher: CallerMatcher<N, A>): CallerArgumentMatcher<N, A> {
  if (!isCallerMatcher(callerMatcher)) {
    throw getInvalidTypeError('Invalid caller matcher supplied to the callerArguments().', {
      expected: ['caller()'],
      received: callerMatcher,
    });
  }
  const args = callerMatcher.metadata.options.args;
  const validateArguments = args
    ? (callArgs: Array<any>) => {
        if (!callArgs || callArgs.length !== args.length) return false;
        return args.some((argValidator, argIndex) => argValidator(callArgs[argIndex]));
      }
    : () => true;
  const matcher = createMatcher<Array<any>, CallerMatcher<N, A>>(
    'callerArguments',
    validateArguments,
    callerMatcher,
  );
  matcher.metadata.type = callerArguments;
  return matcher;
}

export function isCallerArgumentMatcher(value: any): value is CallerArgumentMatcher<any, any> {
  return isMatcher(value) && value.metadata.type === callerArguments;
}
