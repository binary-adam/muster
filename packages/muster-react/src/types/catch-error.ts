import {
  createMatcher,
  ErrorFallbackGenerator,
  getInvalidTypeError,
  isMatcher,
  Matcher,
  NodeDefinition,
  NodeLike,
  types,
} from '@dws/muster';
import { Props } from '../container-types';
import { sanitizeMatcher } from '../utils/sanitize-matcher';
import { sanitizeProps } from '../utils/sanitize-props';
import { isInjectedMatcher } from './injected';
import { TreeMatcher } from './tree';

export interface CatchErrorMatcherOptions<T, O, M extends Matcher<T, O | any>> {
  fallback: ErrorFallbackGenerator;
  type: M | TreeMatcher<T>;
}

export type CatchErrorMatcher<T, O, M extends Matcher<T, O | any>> = Matcher<
  T,
  CatchErrorMatcherOptions<T, O, M>
>;

/**
 * Creates a matcher used as part of the `props` section of the [[container]] and [[simpleContainer]]. This matcher
 * informs muster-react that a given property can sometimes resolve to an [[error]] node, and that Muster should
 * replace that error with a `fallback` specified by the [[catchError]] prop type. The `fallback` can be defined as a
 * pure value, Muster [[NodeDefinition]] or a fallback generator function. The fallback generator function is a function
 * that gets called when an error is encountered, and is expected to return a [[NodeDefinition]] that can be used
 * as a fallback.
 *
 *
 * @example **Replacing an error with a string**
 * ```js
 * import { container, error, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     name: error('Some error'),
 *   },
 *   props: {
 *     name: propTypes.catchError('Something went wrong', true),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ name }) => <h1>{name}</h1>);
 * // Rendered JSX:
 * // <h1>Something went wrong</h1>
 * ```
 * This example shows how to handle graph errors by replacing the error with a fallback value (in this case a string).
 * If the `name` prop was changed to `name: true` the component wouldn't render at all.
 *
 *
 * @example **Replacing a branch error with a string**
 * ```js
 * import { container, error, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     user: {
 *       firstName: 'Bob',
 *       lastName: error('Some error'),
 *     },
 *   },
 *   props: {
 *     user: propTypes.catchError('Something went wrong', {
 *       firstName: true,
 *       lastName: true,
 *     }),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ user }) => {
 *   if (typeof user === 'string') return <h1>Error: {user}</h1>;
 *   return <p>{user.firstName} {user.lastName}</p>;
 * });
 * // Rendered JSX:
 * // <h1>Error: Something went wrong</h1>
 * ```
 * This example shows how to handle branch-level graph errors by replacing the branch that returned an error with a string.
 * In both examples the replacement value was a string, but it doesn't have to be.
 *
 *
 * @example **Replacing a branch error with an object**
 * ```js
 * import { container, error, propTypes } from '@dws/muster-react';
 *
 * const myContainer = container({
 *   graph: {
 *     user: {
 *       firstName: 'Bob',
 *       lastName: error('Some error'),
 *     },
 *   },
 *   props: {
 *     user: propTypes.catchError({ error: 'Custom error' }, {
 *       firstName: true,
 *       lastName: true,
 *     }),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ user }) => {
 *   if (user.error) return <h1>Error: {user.error}</h1>;
 *   return <p>{user.firstName} {user.lastName}</p>;
 * });
 * // Rendered JSX:
 * // <h1>Error: Custom error</h1>
 * ```
 * This example shows how to handle branch-level graph errors by replacing the branch with an object. The fallback object
 * doesn't have to be of the same shape as the expected branch.
 */
export function catchError<T, O, M extends Matcher<T, O>>(
  fallback: ErrorFallbackGenerator | NodeDefinition | NodeLike,
  type: true | M | Props<T>,
): CatchErrorMatcher<T, O | any, M | TreeMatcher<T>> {
  if (
    (isMatcher(type) && (isInjectedMatcher(type) || isCatchErrorMatcher(type))) ||
    (!isMatcher(type) &&
      ((typeof type === 'boolean' && !type) ||
        ((typeof type !== 'object' || type === null) && typeof type !== 'boolean')))
  ) {
    throw getInvalidTypeError('Invalid parameters supplied to the catchError().', {
      expected: ['Matcher<any, any>', 'getter()', 'list()', 'tree()', '{...}', 'true'],
      received: type,
    });
  }
  const sanitizedMatcher: M | TreeMatcher<T> =
    isMatcher(type) || typeof type === 'boolean'
      ? (sanitizeMatcher(typeof type === 'boolean' ? types.any : type) as M)
      : (sanitizeProps(type) as TreeMatcher<T>);
  const options: CatchErrorMatcherOptions<T, O, M> = {
    fallback,
    type: sanitizedMatcher,
  };
  const matcher = createMatcher<T, CatchErrorMatcherOptions<T, O | any, M | TreeMatcher<T>>>(
    'catchError',
    (value: any) => sanitizedMatcher(value) || value === undefined || value === null,
    options,
  );
  matcher.metadata.type = catchError as any;
  return matcher;
}

export function isCatchErrorMatcher(
  value: any,
): value is CatchErrorMatcher<any, any, Matcher<any, any>> {
  return isMatcher(value) && value.metadata.type === (catchError as any);
}
