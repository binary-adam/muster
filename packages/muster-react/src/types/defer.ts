import {
  createMatcher,
  DeferNodeFallbackGenerator,
  getInvalidTypeError,
  isMatcher,
  Matcher,
  NodeDefinition,
  NodeLike,
  types,
} from '@dws/muster';
import { Props } from '../container-types';
import { sanitizeMatcher } from '../utils/sanitize-matcher';
import { sanitizeProps } from '../utils/sanitize-props';
import { GetterMatcher } from './getter';
import { isInjectedMatcher } from './injected';
import { TreeMatcher } from './tree';

export interface DeferMatcherOptions<T, O, M extends Matcher<T, O | any>> {
  fallback: DeferNodeFallbackGenerator | undefined;
  type: M | TreeMatcher<T>;
}

export type DeferMatcher<T, O, M extends Matcher<T, O | any>> = Matcher<
  T,
  DeferMatcherOptions<T, O, M>
>;

/**
 * Creates a matcher used as part of the `props` section of the [[container]] and [[simpleContainer]].
 * Normally the Muster React component waits for all requested properties to resolve to a non-[[pending]] nodes.
 * This prevents component from being rendered in an inconsistent state.
 * However, this matcher informs muster-react that a given property can sometimes resolve to a [[pending]] node,
 * and when that happens Muster should replace the [[pending]] value with either a `fallback` value, or with `null` when
 * no `fallback` is given.
 * The `fallback` can be defined as an JS value, a Muster node, or a fallback generator function. The fallback generator
 * function receives a previous value of deferred node, or undefined if there was no previous value, and is expected
 * to return a Muster node.
 *
 *
 * @example **Deferring a leaf**
 * ```js
 * import { container, fromPromise, propTypes } from '@dws/muster-react';
 *
 * let resolvePromise;
 *
 * const myContainer = container({
 *   graph: {
 *     firstName: fromPromise(() =>
 *       new Promise((resolve) => { resolvePromise = resolve; }),
 *     ),
 *   },
 *   props: {
 *     firstName: propTypes.defer(true),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ firstName }) => <h1>{firstName || 'Loading...'}</h1>);
 *
 * // Rendered JSX:
 * // <h1>Loading...</h1>
 * ```
 * This example shows how to use the [[defer]] prop type to instruct muster-react not to wait for a given property
 * to load before rendering the component. During the first render the `firstName` property is equal to `null`. Calling
 * the `resolvePromise` with a value will cause a new render of this component with a `firstName` equal to that value.
 *
 *
 * @example **Deferring a branch**
 * ```js
 * import { container, fromPromise, propTypes } from '@dws/muster-react';
 *
 * let resolvePromise;
 *
 * const myContainer = container({
 *   graph: {
 *     user: fromPromise(() =>
 *       new Promise((resolve) => { resolvePromise = resolve; }),
 *     ),
 *   },
 *   props: {
 *     user: propTypes.defer({
 *       firstName: true,
 *       lastName: true,
 *     }),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ user }) => {
 *   if (!user) return <h1>Loading...</h1>;
 *   return <h1>Hello, {user.firstName} {user.lastName}</h1>;
 * });
 *
 * // Rendered JSX:
 * // <h1>Loading...</h1>
 * ```
 * This example show that the [[defer]] can also be used to defer branches.
 *
 *
 * @example **Providing a static fallback value**
 * ```js
 * import { container, fromPromise, propTypes } from '@dws/muster-react';
 *
 * let resolvePromise;
 *
 * const myContainer = container({
 *   graph: {
 *     firstName: fromPromise(() =>
 *       new Promise((resolve) => { resolvePromise = resolve; }),
 *     ),
 *   },
 *   props: {
 *     firstName: propTypes.defer('Loading first name', true),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ firstName }) => <h1>{firstName}</h1>);
 *
 * // Rendered JSX:
 * // <h1>Loading first name</h1>
 * ```
 * This example shows how to use the [[defer]] to provide a primitive fallback value for a deferred property.
 *
 *
 * @example **Defer a list**
 * ```js
 * import { container, fromPromise, propTypes } from '@dws/muster-react';
 *
 * let resolvePromise;
 *
 * const myContainer = container({
 *   graph: {
 *     friends: fromPromise(() =>
 *       new Promise((resolve) => { resolvePromise = resolve; }),
 *     ),
 *   },
 *   props: {
 *     friends: propTypes.defer('Loading friends...', propTypes.list({
 *       id: true,
 *       firstName: true,
 *       lastName: true,
 *     })),
 *   },
 * });
 *
 * const MyComponent = myContainer(({ friends }) => {
 *   if (!Array.isArray(friends)) return <h1>{friends}</h1>;
 *   return (
 *     <ul>
 *       {friends.map((friend) =>
 *         <li key={friend.id}>{friend.firstName} {friend.lastName}</li>
 *       )}
 *     </ul>
 *   )
 * });
 *
 * // Rendered JSX:
 * // <h1>Loading friends...</h1>
 * ```
 * This example shows that the [[defer]] can also be used to defer lists. The fallback value is defined as string, but
 * could also be defined as an empty array, or as any object you want.
 */
export function defer<T, O, M extends Matcher<T, O>>(
  fallback: DeferNodeFallbackGenerator | NodeDefinition | NodeLike,
  type: M,
): DeferMatcher<T, O, M>;
export function defer<T, O, M extends Matcher<T, O>>(type: M): DeferMatcher<T, O, M>;
export function defer<T>(
  fallback: DeferNodeFallbackGenerator | NodeDefinition | NodeLike,
  props: Props<T>,
): DeferMatcher<T, any, TreeMatcher<T>>;
export function defer<T>(props: Props<T>): DeferMatcher<T, any, TreeMatcher<T>>;
export function defer(
  fallback: DeferNodeFallbackGenerator | NodeDefinition | NodeLike,
  prop: true,
): DeferMatcher<any, any, GetterMatcher<any, any, any>>;
export function defer(prop: true): DeferMatcher<any, any, GetterMatcher<any, any, any>>;
export function defer<T, O, M extends Matcher<T, O>>(
  ...args: Array<M | Props<T> | DeferNodeFallbackGenerator | NodeDefinition | NodeLike | true>
): DeferMatcher<T, O | any, M | TreeMatcher<T> | GetterMatcher<any, any, any>> {
  const type = args.length === 1 ? args[0] : args[1];
  const fallback = args.length === 2 ? args[0] : undefined;
  if (
    (isMatcher(type) && (isInjectedMatcher(type) || isDeferMatcher(type))) ||
    (!isMatcher(type) &&
      (typeof type !== 'object' || type === null) &&
      (typeof type !== 'boolean' || type === false))
  ) {
    throw getInvalidTypeError('Invalid parameters supplied to the defer().', {
      expected: ['Matcher<any, any>', 'getter()', 'list()', 'tree()', '{...}', 'true'],
      received: type,
    });
  }
  const sanitizedMatcher: M | TreeMatcher<T> =
    isMatcher(type) || typeof type === 'boolean'
      ? (sanitizeMatcher(typeof type === 'boolean' ? types.any : type) as M)
      : (sanitizeProps(type) as TreeMatcher<T>);
  const options: DeferMatcherOptions<T, O, M> = {
    fallback,
    type: sanitizedMatcher,
  };
  const matcher = createMatcher<T, DeferMatcherOptions<T, O, M>>(
    'defer',
    (value: any) => sanitizedMatcher(value) || value === undefined || value === null,
    options,
  );
  matcher.metadata.type = defer;
  return matcher;
}

export function isDeferMatcher(value: any): value is DeferMatcher<any, any, Matcher<any, any>> {
  return isMatcher(value) && value.metadata.type === defer;
}
