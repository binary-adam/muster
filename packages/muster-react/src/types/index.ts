import {
  // @ts-ignore This is just to shut TypeScript up about unused parameter
  // which is actually used by export const propTypes = { ... };
  Matcher,
  types as musterTypes,
} from '@dws/muster';
import { caller, callerArguments } from './caller';
import { catchError } from './catch-error';
import { defer } from './defer';
import { getter } from './getter';
import { injected } from './injected';
import { isLoading } from './is-loading';
import { list } from './list';
import { setter, setterValue } from './setter';
import { tree } from './tree';

export const propTypes = {
  ...musterTypes,
  caller,
  callerArguments,
  catchError,
  defer,
  getter,
  injected,
  isLoading,
  list,
  setter,
  setterValue,
  tree,
};

export {
  CallerArgumentMatcher,
  CallerMatcher,
  CallerOptions,
  isCallerArgumentMatcher,
  isCallerMatcher,
} from './caller';
export { CatchErrorMatcher, CatchErrorMatcherOptions, isCatchErrorMatcher } from './catch-error';
export { DeferMatcher, isDeferMatcher } from './defer';
export { GetterMatcher, GetterOptions, isGetterMatcher } from './getter';
export { InjectedMatcher, InjectedOptions, isInjectedMatcher } from './injected';
export { IsLoadingMatcher, isIsLoadingMatcher } from './is-loading';
export { isListMatcher, ListMatcher, ListOptions } from './list';
export {
  isSetterMatcher,
  isSetterValueMatcher,
  SetterMatcher,
  SetterOptions,
  SetterValueMatcher,
} from './setter';
export { isTreeMatcher, TreeFields, TreeMatcher } from './tree';
