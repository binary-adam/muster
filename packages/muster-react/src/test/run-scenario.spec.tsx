import muster, { NodeDefinition, ref, set, types, variable } from '@dws/muster';
import * as React from 'react';
import { operation, runScenario } from '.';
import { global } from '..';
import { container } from '../container';
import { propTypes } from '../types';

(() => {
  const globalMuster = () => muster({ outer: variable('foo') });

  const ExampleContainer = container({
    graph: {
      inner: ref(global('outer')),
    },
    props: {
      inner: types.any,
      setInner: propTypes.setter('inner', types.string),
    },
  });

  interface ExampleComponentProps {
    inner: string;
    setInner: (value: any) => Promise<NodeDefinition>;
  }

  // tslint:disable-next-line:function-name
  function ExampleComponent(props: ExampleComponentProps) {
    return (
      <div>
        <InnerValueComponent value={props.inner} />
        <button id="submit" onClick={() => props.setInner('clicked')}>
          Update
        </button>
      </div>
    );
  }

  interface InnerValueComponentProps {
    value: string;
  }

  // tslint:disable-next-line:function-name
  function InnerValueComponent(props: InnerValueComponentProps) {
    return <div>{props.value}</div>;
  }

  runScenario({
    description: 'Nested component (shallow)',
    graph: globalMuster,
    container: ExampleContainer,
    component: ExampleComponent,
    shallow: true,
    expected: {
      snapshot: true,
      props: {
        inner: 'foo',
        setInner: expect.any(Function),
      },
    },
    assert(results) {
      expect(
        results.view
          .children()
          .first()
          .prop('value'),
      ).toBe('foo');
    },
    operations: [
      operation({
        description: 'AND a variable is updated in the global graph',
        input: set(ref('outer'), 'bar'),
        expected: {
          snapshot: true,
          props: {
            inner: 'bar',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('bar');
        },
      }),
      operation({
        description: 'AND a variable is updated via the component DOM',
        input: (results) => results.view.find('#submit').simulate('click'),
        expected: {
          snapshot: true,
          graph: {
            outer: 'clicked',
          },
          props: {
            inner: 'clicked',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('clicked');
        },
      }),
      operation({
        description: 'AND a variable is updated via the component props',
        input: (results) => results.props.setInner('baz'),
        expected: {
          snapshot: true,
          graph: {
            outer: 'baz',
          },
          props: {
            inner: 'baz',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('baz');
        },
      }),
    ],
  });

  runScenario({
    description: 'Nested component (deep)',
    graph: globalMuster,
    container: ExampleContainer,
    component: ExampleComponent,
    expected: {
      snapshot: true,
      props: {
        inner: 'foo',
        setInner: expect.any(Function),
      },
    },
    assert(results) {
      expect(
        results.view
          .children()
          .first()
          .prop('value'),
      ).toBe('foo');
    },
    operations: [
      operation({
        description: 'AND a variable is updated in the global graph',
        input: set(ref('outer'), 'bar'),
        expected: {
          snapshot: true,
          props: {
            inner: 'bar',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('bar');
        },
      }),
      operation({
        description: 'AND a variable is updated via the component DOM',
        input: (results) => results.view.find('#submit').simulate('click'),
        expected: {
          snapshot: true,
          graph: {
            outer: 'clicked',
          },
          props: {
            inner: 'clicked',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('clicked');
        },
      }),
      operation({
        description: 'AND a variable is updated via the component props',
        input: (results) => results.props.setInner('baz'),
        expected: {
          snapshot: true,
          graph: {
            outer: 'baz',
          },
          props: {
            inner: 'baz',
            setInner: expect.any(Function),
          },
        },
        assert(results) {
          expect(
            results.view
              .children()
              .first()
              .prop('value'),
          ).toBe('baz');
        },
      }),
    ],
  });
})();
