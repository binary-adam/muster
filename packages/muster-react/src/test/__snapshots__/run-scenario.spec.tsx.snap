// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested component (deep) WHEN the component is rendered AND a variable is updated in the global graph SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="bar"
  >
    <div>
      bar
    </div>
  </InnerValueComponent>
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (deep) WHEN the component is rendered AND a variable is updated via the component DOM SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="clicked"
  >
    <div>
      clicked
    </div>
  </InnerValueComponent>
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (deep) WHEN the component is rendered AND a variable is updated via the component props SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="baz"
  >
    <div>
      baz
    </div>
  </InnerValueComponent>
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (deep) WHEN the component is rendered SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="foo"
  >
    <div>
      foo
    </div>
  </InnerValueComponent>
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (shallow) WHEN the component is rendered AND a variable is updated in the global graph SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="bar"
  />
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (shallow) WHEN the component is rendered AND a variable is updated via the component DOM SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="clicked"
  />
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (shallow) WHEN the component is rendered AND a variable is updated via the component props SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="baz"
  />
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;

exports[`Nested component (shallow) WHEN the component is rendered SHOULD return the correct values 1`] = `
<div>
  <InnerValueComponent
    value="foo"
  />
  <button
    id="submit"
    onClick={[Function]}
  >
    Update
  </button>
</div>
`;
